### PR TITLE
determine coordinator mode by is_grad

### PR DESCRIPTION
**Pull request by ByronHsu**
### Why?

As described in https://github.com/microsoft/DeepSpeed/issues/4922, the mode of coordinator is determined by `sub_module.training` at pre and 1st post forward hook, but by `is_grad` at the 2nd post forward hook for the case of inference.

If `model.train()` and `torch.no_grad()` are both set, in pre and 1st post forward hook, the state of "train coordinator" is updated, but in 2nd post forward hook, the state of "inference coordinator" is cleared and reset. It means that the resetting is on "eval coordinator", so the state of "train coordinator" will keep accumulating. 

### Test

WIP

### CODE REVIEW

**Review by HeyangQin — APPROVED** [review]

**Review by HeyangQin — CHANGES_REQUESTED** [review]
I validated this PR with https://github.com/microsoft/DeepSpeedExamples/blob/master/applications/DeepSpeed-Chat/training/step3_rlhf_finetuning/training_scripts/llama2/run_llama2_7b.sh. Unfortunately, RLHF training with this PR fails at step 61 due to gradient overflow whereas the current DeepSpeed master trains well. I guess we need a different way to handle the coordinator mode. Will investigate more when I get the bandwidth.

**Comment by loadams** [comment]
@ByronHsu - ping on Heyang's comment if you are able to follow up on that.

**Comment by jomayeri** [comment]
Closing due to lack response.
